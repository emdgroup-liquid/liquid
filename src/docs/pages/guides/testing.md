---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

