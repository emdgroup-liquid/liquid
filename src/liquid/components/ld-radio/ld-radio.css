:host,
.ld-radio {
  /* layout */
  --ld-radio-size: 1.25rem;
  --ld-radio-dot-size: 0.625rem;

  /* colors */
  --ld-radio-bg-col: var(--ld-col-wht);
  --ld-radio-bg-col-hover: var(--ld-col-neutral-010);
  --ld-radio-bg-col-active: var(--ld-col-neutral-050);
  --ld-radio-disabled-col: var(--ld-col-neutral-300);
  --ld-radio-disabled-bg-col: var(--ld-col-neutral-100);
  --ld-radio-dark-bg-col: var(--ld-col-neutral-050);
  --ld-radio-dark-bg-col-hover: var(--ld-col-neutral-100);
  --ld-radio-dark-bg-col-active: var(--ld-col-neutral-300);
  --ld-radio-checked-col: var(--ld-col-wht);
  --ld-radio-checked-col-active: var(--ld-col-wht);
  --ld-radio-invalid-checked-col: var(--ld-col-wht);
  --ld-radio-invalid-checked-col-active: var(--ld-col-wht);

  /* themable colors */
  --ld-radio-col: var(--ld-thm-primary);
  --ld-radio-col-hover: var(--ld-thm-primary-hover);
  --ld-radio-col-active: var(--ld-thm-primary-active);
  --ld-radio-col-focus: var(--ld-thm-primary-focus);
  --ld-radio-invalid-col: var(--ld-thm-error);
  --ld-radio-invalid-col-hover: var(--ld-thm-error-hover);
  --ld-radio-invalid-col-focus: var(--ld-thm-error-focus);
  --ld-radio-invalid-col-active: var(--ld-thm-error-active);
  --ld-radio-warn-col: var(--ld-thm-warning);
  --ld-radio-warn-col-hover: var(--ld-thm-warning-hover);
  --ld-radio-warn-col-focus: var(--ld-thm-warning-focus);
  --ld-radio-warn-col-active: var(--ld-thm-warning-active);
  --ld-radio-warn-checked-col: var(--ld-thm-warning-active);
  --ld-radio-warn-checked-col-active: var(--ld-thm-warning-hover);

  position: relative;
  display: inline-flex;
  flex-shrink: 0;
  width: var(--ld-radio-size);
  height: var(--ld-radio-size);
  min-width: auto !important;
  box-sizing: border-box;

  input {
    appearance: none;
    background-color: transparent; /* Required for Mobile Sarari */
    height: 100%;
    inset: 0;
    margin: 0;
    position: absolute;
    width: 100%;
    z-index: 1;

    &:checked {
      ~ .ld-radio__dot {
        visibility: inherit;
      }
    }

    &:where(:disabled),
    &:where([aria-disabled='true']) {
      ~ .ld-radio__dot {
        background-color: var(--ld-radio-disabled-col);
      }
      ~ .ld-radio__box {
        background-color: transparent;
        box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-radio-disabled-bg-col);
      }

      &:checked {
        ~ .ld-radio__box {
          background-color: var(--ld-radio-disabled-bg-col);
        }
      }
    }

    &:where(:not(:disabled, [aria-disabled='true'])) {
      cursor: pointer;

      ~ .ld-radio__dot {
        background-color: var(--ld-radio-col-active);
      }
      ~ .ld-radio__box {
        box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-radio-col);
      }

      &:focus:focus-visible {
        ~ .ld-radio__box {
          background-color: transparent;
          box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-radio-col-focus);
        }
      }
      &:active,
      &:active:focus-visible {
        ~ .ld-radio__box {
          background-color: var(--ld-radio-bg-col-active);
          box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-radio-col-active);
        }
      }

      &:checked {
        ~ .ld-radio__dot {
          background-color: var(--ld-radio-checked-col);
        }
        ~ .ld-radio__box {
          background-color: var(--ld-radio-col);
          box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-radio-col);
        }

        &:focus:focus-visible {
          ~ .ld-radio__dot {
            background-color: var(--ld-radio-col-active);
          }
          ~ .ld-radio__box {
            background-color: var(--ld-radio-col-focus);
            box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-radio-col-focus);
          }
        }
        &:active,
        &:active:focus-visible {
          ~ .ld-radio__dot {
            background-color: var(--ld-radio-checked-col-active);
          }
          ~ .ld-radio__box {
            background-color: var(--ld-radio-col-active);
            box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-radio-col-active);
          }
        }
      }
    }
  }

  :where(input:not(:disabled, input[aria-disabled='true'])) {
    ~ .ld-radio__box {
      background-color: var(--ld-radio-bg-col);
    }

    @media (hover: hover) {
      &:hover {
        ~ .ld-radio__box {
          background-color: var(--ld-radio-bg-col-hover);
          box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-radio-col-hover);
        }

        &:checked {
          ~ .ld-radio__dot {
            color: var(--ld-radio-checked-col);
          }
          ~ .ld-radio__box {
            background-color: var(--ld-radio-col-hover);
            box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-radio-col-hover);
          }
        }
      }
    }
  }
}

.ld-radio.ld-radio--dark,
:host(.ld-radio.ld-radio--dark) {
  :where(input) {
    &:not(:disabled, input[aria-disabled='true'], :checked) {
      ~ .ld-radio__box {
        background-color: var(--ld-radio-dark-bg-col);
      }

      @media (hover: hover) {
        &:hover {
          ~ .ld-radio__box {
            background-color: var(--ld-radio-dark-bg-col-hover);
          }
        }
      }

      &:active,
      &:active:focus-visible {
        ~ .ld-radio__box {
          background-color: var(--ld-radio-dark-bg-col-active);
        }
      }
    }
  }
}

.ld-radio__dot {
  visibility: hidden;
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
  pointer-events: none;
  z-index: 1;
  width: var(--ld-radio-dot-size);
  height: var(--ld-radio-dot-size);
}

.ld-radio__box {
  position: absolute;
  width: 100%;
  height: 100%;
  pointer-events: none;
  box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-col-rblck4);
}

:host,
.ld-radio,
.ld-radio__dot,
.ld-radio__box {
  border-radius: var(--ld-br-full);
}

:host(.ld-radio--highlight),
.ld-radio--highlight {
  --ld-radio-col: var(--ld-radio-warn-col);
  --ld-radio-col-hover: var(--ld-radio-warn-col-hover);
  --ld-radio-col-active: var(--ld-radio-warn-col-focus);
  --ld-radio-col-focus: var(--ld-radio-warn-col-active);
  --ld-radio-checked-col: var(--ld-radio-warn-checked-col);
  --ld-radio-checked-col-active: var(--ld-radio-warn-checked-col-active);
}

:host(.ld-radio--invalid),
:host(.ld-radio--danger),
.ld-radio--invalid,
.ld-radio--danger {
  --ld-radio-col: var(--ld-radio-invalid-col);
  --ld-radio-col-hover: var(--ld-radio-invalid-col-hover);
  --ld-radio-col-active: var(--ld-radio-invalid-col-focus);
  --ld-radio-col-focus: var(--ld-radio-invalid-col-active);
  --ld-radio-checked-col: var(--ld-radio-invalid-checked-col);
  --ld-radio-checked-col-active: var(--ld-radio-invalid-checked-col-active);
}
