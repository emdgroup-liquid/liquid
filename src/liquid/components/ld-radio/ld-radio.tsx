import {
  Component,
  Element,
  h,
  Host,
  Method,
  Prop,
  State,
  Watch,
} from '@stencil/core'
import { cloneAttributes } from '../../utils/cloneAttributes'
import { getClassNames } from '../../utils/getClassNames'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part input - Actual input element
 */
@Component({
  tag: 'ld-radio',
  styleUrl: 'ld-radio.css',
  shadow: true,
})
export class LdRadio implements InnerFocusable, ClonesAttributes {
  @Element() el: HTMLInputElement

  private attributesObserver: MutationObserver

  private input: HTMLInputElement
  private hiddenInput: HTMLInputElement

  /** Automatically focus the form control when the page is loaded. */
  @Prop() autofocus?: boolean

  /** Indicates whether the radio button is selected. */
  @Prop({ mutable: true }) checked = false

  /** Disabled state of the radio. */
  @Prop() disabled: boolean

  /** Associates the control with a form element. */
  @Prop() form?: string

  /** Set this property to `true` in order to mark the radio visually as invalid. */
  @Prop() invalid: boolean

  /** Display mode. */
  @Prop() mode?: 'highlight' | 'danger'

  /** Used to specify the name of the control. */
  @Prop() name: string

  /** The value is not editable. */
  @Prop() readonly?: boolean

  /** Set this property to `true` in order to mark the checkbox as required. */
  @Prop() required: boolean

  /** radio tone. Use `'dark'` on white backgrounds. Default is a light tone. */
  @Prop() tone: 'dark'

  /** The input value. */
  @Prop() value: string

  @State() clonedAttributes

  /** Sets focus on the radio button. */
  @Method()
  async focusInner() {
    if (this.input !== undefined) {
      this.input.focus()
    }
  }

  @Watch('checked')
  @Watch('name')
  @Watch('value')
  updateHiddenInput() {
    const outerForm = this.el.closest('form')
    if (!this.hiddenInput && this.name && (outerForm || this.form)) {
      this.createHiddenInput()
    }

    if (this.hiddenInput) {
      if (!this.name) {
        this.hiddenInput.remove()
        this.hiddenInput = undefined
        return
      }

      this.hiddenInput.name = this.name
      this.hiddenInput.checked = this.checked

      if (this.value) {
        this.hiddenInput.value = this.value
      } else {
        this.hiddenInput.removeAttribute('value')
      }

      if (this.form) {
        this.hiddenInput.setAttribute('form', this.form)
      } else if (this.hiddenInput.getAttribute('form')) {
        if (outerForm) {
          this.hiddenInput.removeAttribute('form')
        } else {
          this.hiddenInput.remove()
          this.hiddenInput = undefined
        }
      }
    }
  }

  private createHiddenInput() {
    this.hiddenInput = document.createElement('input')
    this.hiddenInput.type = 'radio'
    this.hiddenInput.style.visibility = 'hidden'
    this.hiddenInput.style.position = 'absolute'
    this.hiddenInput.style.pointerEvents = 'none'
    this.el.appendChild(this.hiddenInput)
  }

  private handleKeyDown = (ev: KeyboardEvent) => {
    switch (ev.key) {
      case 'ArrowUp':
      case 'ArrowLeft': {
        ev.preventDefault()
        this.focusAndSelect('prev')
        return
      }
      case 'ArrowDown':
      case 'ArrowRight': {
        ev.preventDefault()
        this.focusAndSelect('next')
        return
      }
    }
  }

  private handleBlur = (ev: FocusEvent) => {
    setTimeout(() => {
      this.el.dispatchEvent(ev)
    })
  }

  private handleFocus = (ev: FocusEvent) => {
    setTimeout(() => {
      this.el.dispatchEvent(ev)
    })
  }

  private handleClick = (ev?: MouseEvent) => {
    if (this.disabled || this.el.getAttribute('aria-disabled') === 'true') {
      ev?.preventDefault()
      return
    }

    if (this.checked) return

    // Uncheck radios with same name.
    if (this.name) {
      // Attribute selector fails in test env, hance filtering with js below.
      Array.from(document.querySelectorAll('ld-radio'))
        .filter((ldRadio) => ldRadio.getAttribute('name') === this.name)
        .forEach((ldRadio) => {
          ldRadio.checked = false
        })
    }

    this.checked = true

    if (!ev.isTrusted) {
      // This happens, when a click event is dispatched on the host element
      // from the outside i.e. on click on a parent ld-label element.
      this.el.dispatchEvent(
        new Event('input', { bubbles: true, composed: true })
      )
    }
  }

  private focusAndSelect(dir: 'next' | 'prev') {
    const ldRadios = Array.from(document.querySelectorAll('ld-radio')).filter(
      (ldRadio) => ldRadio.getAttribute('name') === this.name
    )
    ldRadios.forEach((ldRadio, index) => {
      if (ldRadio === (this.el as unknown as HTMLLdRadioElement)) {
        const targetLdRadio = ldRadios[index + (dir === 'next' ? 1 : -1)]
        if (targetLdRadio) {
          targetLdRadio.focusInner()
          targetLdRadio.click()
        }
      }
    })
  }

  componentWillLoad() {
    this.attributesObserver = cloneAttributes.call(this, ['tone', 'mode'])

    const outerForm = this.el.closest('form')

    if (this.name && (outerForm || this.form)) {
      this.createHiddenInput()
      this.hiddenInput.checked = this.checked
      this.hiddenInput.name = this.name

      if (this.form) {
        this.hiddenInput.setAttribute('form', this.form)
      }

      if (this.value) {
        this.hiddenInput.value = this.value
      }
    }
  }

  componentDidLoad() {
    if (this.autofocus) {
      this.focusInner()
    }
  }

  disconnectedCallback() {
    this.attributesObserver?.disconnect()
  }

  render() {
    const cl = [
      'ld-radio',
      this.mode && `ld-radio--${this.mode}`,
      this.tone && `ld-radio--${this.tone}`,
      this.invalid && 'ld-radio--invalid',
    ]

    return (
      <Host part="root" class={getClassNames(cl)} onClick={this.handleClick}>
        <input
          {...this.clonedAttributes}
          part="input focusable"
          onBlur={this.handleBlur}
          onFocus={this.handleFocus}
          onKeyDown={this.handleKeyDown}
          ref={(ref) => (this.input = ref)}
          type="radio"
          disabled={this.disabled}
          checked={this.checked}
          tabIndex={
            this.checked
              ? parseInt(this.el.getAttribute('tabindex')) || undefined
              : -1
          }
          value={this.value}
        />
        <div part="dot" class="ld-radio__dot"></div>
        <div class="ld-radio__box" part="box"></div>
      </Host>
    )
  }
}
