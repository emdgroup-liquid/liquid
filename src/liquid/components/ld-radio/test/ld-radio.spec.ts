import { newSpecPage } from '@stencil/core/testing'
import { LdRadio } from '../ld-radio'

describe('ld-radio', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [LdRadio],
      html: `<ld-radio></ld-radio>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('disabled', async () => {
    const page = await newSpecPage({
      components: [LdRadio],
      html: `<ld-radio disabled></ld-radio>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('heighlight', async () => {
    const page = await newSpecPage({
      components: [LdRadio],
      html: `<ld-radio mode="highlight"></ld-radio>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('danger', async () => {
    const page = await newSpecPage({
      components: [LdRadio],
      html: `<ld-radio mode="danger"></ld-radio>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('tone', async () => {
    const page = await newSpecPage({
      components: [LdRadio],
      html: `<ld-radio tone="dark"></ld-radio>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('updates checked prop on checked value change', async () => {
    const page = await newSpecPage({
      components: [LdRadio],
      html: `<ld-radio></ld-radio>`,
    })
    const ldRadio = page.root
    expect(ldRadio.checked).toBe(undefined)

    const input = ldRadio.shadowRoot.querySelector('input')
    expect(input.checked).toBe(false)

    ldRadio.click()
    await page.waitForChanges()

    expect(ldRadio.getAttribute('checked')).not.toBe(null)
    expect(input.checked).toBe(true)
  })

  it('emits focus and blur event', async () => {
    const page = await newSpecPage({
      components: [LdRadio],
      html: `<ld-radio></ld-radio>`,
    })
    const ldRadio = page.root
    const input = ldRadio.shadowRoot.querySelector('input')

    const handlers = {
      onFocus() {
        return
      },
      onBlur() {
        return
      },
    }

    const spyFocus = jest.spyOn(handlers, 'onFocus')
    ldRadio.addEventListener('focus', handlers.onFocus)
    input.dispatchEvent(new Event('focus'))
    jest.advanceTimersByTime(0)
    expect(spyFocus).toHaveBeenCalled()

    const spyBlur = jest.spyOn(handlers, 'onBlur')
    ldRadio.addEventListener('blur', handlers.onBlur)
    input.dispatchEvent(new Event('blur'))
    jest.advanceTimersByTime(0)
    expect(spyBlur).toHaveBeenCalled()
  })

  it('emits input event on click', async () => {
    const page = await newSpecPage({
      components: [LdRadio],
      html: `<ld-radio />`,
    })
    const ldRadio = page.root

    const spyInput = jest.fn()
    ldRadio.addEventListener('input', spyInput)
    ldRadio.click()

    expect(spyInput).toHaveBeenCalled()
  })

  it('allows to set inner focus', async () => {
    const page = await newSpecPage({
      components: [LdRadio],
      html: `<ld-radio />`,
    })
    const ldRadio = page.root
    const input = ldRadio.shadowRoot.querySelector('input')

    input.focus = jest.fn()
    await ldRadio.focusInner()

    expect(input.focus).toHaveBeenCalled()
  })

  it('unchecks radios with same name when checked', async () => {
    const page = await newSpecPage({
      components: [LdRadio],
      html: `
        <ld-radio name="foo" />
        <ld-radio name="bar" />
        <ld-radio name="foo" checked />
        <ld-radio name="baz" checked />
      `,
    })

    const ldRadios = Array.from(page.body.querySelectorAll('ld-radio'))

    expect(ldRadios[0].getAttribute('checked')).toBe(null)
    expect(ldRadios[1].getAttribute('checked')).toBe(null)
    expect(ldRadios[2].getAttribute('checked')).not.toBe(null)
    expect(ldRadios[3].getAttribute('checked')).not.toBe(null)

    ldRadios[0].click()
    await page.waitForChanges()

    expect(ldRadios[0].getAttribute('checked')).not.toBe(null)
    expect(ldRadios[1].getAttribute('checked')).toBe(null)
    expect(ldRadios[2].getAttribute('checked')).toBe(null)
    expect(ldRadios[3].getAttribute('checked')).not.toBe(null)
  })

  it('moves focus and selection', async () => {
    const page = await newSpecPage({
      components: [LdRadio],
      html: `
        <ld-radio name="foo" checked />
        <ld-radio name="bar" />
        <ld-radio name="foo" />
        <ld-radio name="baz" />
        <ld-radio name="foo" />
      `,
    })

    const ldRadios = Array.from(page.body.querySelectorAll('ld-radio'))
    const radios = ldRadios.map((ldRadio) =>
      ldRadio.shadowRoot.querySelector('input')
    )
    ldRadios.forEach((ldRadio) => {
      ldRadio.click = jest.fn()
    })
    radios.forEach((radio) => {
      radio.focus = jest.fn()
    })

    radios[0].dispatchEvent(
      new KeyboardEvent('keydown', { key: 'Escape', bubbles: true })
    ) // nothing shall happen
    await page.waitForChanges()

    expect(ldRadios[0].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[1].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[2].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[3].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[4].click).toHaveBeenCalledTimes(0)

    expect(radios[0].focus).toHaveBeenCalledTimes(0)
    expect(radios[1].focus).toHaveBeenCalledTimes(0)
    expect(radios[2].focus).toHaveBeenCalledTimes(0)
    expect(radios[3].focus).toHaveBeenCalledTimes(0)
    expect(radios[4].focus).toHaveBeenCalledTimes(0)

    radios[0].dispatchEvent(
      new KeyboardEvent('keydown', { key: 'ArrowRight', bubbles: true })
    )
    await page.waitForChanges()

    expect(ldRadios[0].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[1].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[2].click).toHaveBeenCalledTimes(1)
    expect(ldRadios[3].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[4].click).toHaveBeenCalledTimes(0)

    expect(radios[0].focus).toHaveBeenCalledTimes(0)
    expect(radios[1].focus).toHaveBeenCalledTimes(0)
    expect(radios[2].focus).toHaveBeenCalledTimes(1)
    expect(radios[3].focus).toHaveBeenCalledTimes(0)
    expect(radios[4].focus).toHaveBeenCalledTimes(0)

    radios[2].dispatchEvent(
      new KeyboardEvent('keydown', { key: 'ArrowDown', bubbles: true })
    )
    await page.waitForChanges()

    expect(ldRadios[0].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[1].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[2].click).toHaveBeenCalledTimes(1)
    expect(ldRadios[3].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[4].click).toHaveBeenCalledTimes(1)

    expect(radios[0].focus).toHaveBeenCalledTimes(0)
    expect(radios[1].focus).toHaveBeenCalledTimes(0)
    expect(radios[2].focus).toHaveBeenCalledTimes(1)
    expect(radios[3].focus).toHaveBeenCalledTimes(0)
    expect(radios[4].focus).toHaveBeenCalledTimes(1)

    radios[4].dispatchEvent(
      new KeyboardEvent('keydown', { key: 'ArrowDown', bubbles: true })
    )
    await page.waitForChanges()

    expect(ldRadios[0].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[1].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[2].click).toHaveBeenCalledTimes(1)
    expect(ldRadios[3].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[4].click).toHaveBeenCalledTimes(1)

    expect(radios[0].focus).toHaveBeenCalledTimes(0)
    expect(radios[1].focus).toHaveBeenCalledTimes(0)
    expect(radios[2].focus).toHaveBeenCalledTimes(1)
    expect(radios[3].focus).toHaveBeenCalledTimes(0)
    expect(radios[4].focus).toHaveBeenCalledTimes(1)

    radios[4].dispatchEvent(
      new KeyboardEvent('keydown', { key: 'ArrowLeft', bubbles: true })
    )
    await page.waitForChanges()

    expect(ldRadios[0].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[1].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[2].click).toHaveBeenCalledTimes(2)
    expect(ldRadios[3].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[4].click).toHaveBeenCalledTimes(1)

    expect(radios[0].focus).toHaveBeenCalledTimes(0)
    expect(radios[1].focus).toHaveBeenCalledTimes(0)
    expect(radios[2].focus).toHaveBeenCalledTimes(2)
    expect(radios[3].focus).toHaveBeenCalledTimes(0)
    expect(radios[4].focus).toHaveBeenCalledTimes(1)

    radios[2].dispatchEvent(
      new KeyboardEvent('keydown', { key: 'ArrowUp', bubbles: true })
    )
    await page.waitForChanges()

    expect(ldRadios[0].click).toHaveBeenCalledTimes(1)
    expect(ldRadios[1].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[2].click).toHaveBeenCalledTimes(2)
    expect(ldRadios[3].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[4].click).toHaveBeenCalledTimes(1)

    expect(radios[0].focus).toHaveBeenCalledTimes(1)
    expect(radios[1].focus).toHaveBeenCalledTimes(0)
    expect(radios[2].focus).toHaveBeenCalledTimes(2)
    expect(radios[3].focus).toHaveBeenCalledTimes(0)
    expect(radios[4].focus).toHaveBeenCalledTimes(1)

    radios[0].dispatchEvent(
      new KeyboardEvent('keydown', { key: 'ArrowLeft', bubbles: true })
    )
    await page.waitForChanges()

    expect(ldRadios[0].click).toHaveBeenCalledTimes(1)
    expect(ldRadios[1].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[2].click).toHaveBeenCalledTimes(2)
    expect(ldRadios[3].click).toHaveBeenCalledTimes(0)
    expect(ldRadios[4].click).toHaveBeenCalledTimes(1)

    expect(radios[0].focus).toHaveBeenCalledTimes(1)
    expect(radios[1].focus).toHaveBeenCalledTimes(0)
    expect(radios[2].focus).toHaveBeenCalledTimes(2)
    expect(radios[3].focus).toHaveBeenCalledTimes(0)
    expect(radios[4].focus).toHaveBeenCalledTimes(1)
  })

  it('creates hidden input field, if inside a form', async () => {
    const page = await newSpecPage({
      components: [LdRadio],
      html: `<form><ld-radio name="example" /></form>`,
    })
    const ldRadio = page.root
    expect(ldRadio).toMatchSnapshot()
  })

  it('sets initial state on hidden input', async () => {
    const page = await newSpecPage({
      components: [LdRadio],
      html: `<form><ld-radio name="example" checked value="test" /></form>`,
    })
    const ldRadio = page.root
    expect(ldRadio.querySelector('input')).toHaveProperty('checked', true)
    expect(ldRadio.querySelector('input')).toHaveProperty('name', 'example')
    expect(ldRadio.querySelector('input')).toHaveProperty('value', 'test')
  })

  it('updates hidden input field', async () => {
    const { root, waitForChanges } = await newSpecPage({
      components: [LdRadio],
      html: `<form><ld-radio name="example" /></form>`,
    })
    const ldRadio = root
    await waitForChanges()

    expect(ldRadio.querySelector('input')).toHaveProperty('name', 'example')
    expect(ldRadio.querySelector('input')).toHaveProperty('checked', false)

    ldRadio.setAttribute('name', 'test')
    await waitForChanges()

    expect(ldRadio.querySelector('input')).toHaveProperty('name', 'test')

    ldRadio.removeAttribute('name')
    ldRadio.name = undefined
    await waitForChanges()

    expect(ldRadio.querySelector('input')).toEqual(null)

    ldRadio.setAttribute('name', 'test')
    await waitForChanges()

    expect(ldRadio.querySelector('input')).toHaveProperty('name', 'test')

    ldRadio.dispatchEvent(new Event('click'))
    await waitForChanges()

    expect(ldRadio.querySelector('input')).toHaveProperty('checked', true)

    ldRadio.setAttribute('value', 'test')
    await waitForChanges()

    expect(ldRadio.querySelector('input')).toHaveProperty('value', 'test')

    ldRadio.removeAttribute('value')
    ldRadio.value = undefined
    await waitForChanges()

    expect(ldRadio.querySelector('input').getAttribute('value')).toEqual(null)
  })

  it('uses hidden input field with referenced form', async () => {
    const { root, waitForChanges } = await newSpecPage({
      components: [LdRadio],
      html: '<ld-radio name="example" form="yolo" />',
    })
    const ldRadio = root
    await waitForChanges()

    expect(ldRadio.querySelector('input')).toHaveProperty('name', 'example')
  })
})
