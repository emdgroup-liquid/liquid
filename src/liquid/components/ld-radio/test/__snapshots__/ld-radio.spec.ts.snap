// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-radio creates hidden input field, if inside a form 1`] = `
<ld-radio class="ld-radio" part="root">
  <mock:shadow-root>
    <input part="input focusable" tabindex="-1" type="radio">
    <div class="ld-radio__dot" part="dot"></div>
    <div class="ld-radio__box" part="box"></div>
  </mock:shadow-root>
  <input type="radio" style="visibility: hidden; position: absolute; pointer-events: none;">
</ld-radio>
`;

exports[`ld-radio danger 1`] = `
<ld-radio class="ld-radio ld-radio--danger" mode="danger" part="root">
  <mock:shadow-root>
    <input part="input focusable" tabindex="-1" type="radio">
    <div class="ld-radio__dot" part="dot"></div>
    <div class="ld-radio__box" part="box"></div>
  </mock:shadow-root>
</ld-radio>
`;

exports[`ld-radio disabled 1`] = `
<ld-radio class="ld-radio" disabled="" part="root">
  <mock:shadow-root>
    <input disabled="" part="input focusable" tabindex="-1" type="radio">
    <div class="ld-radio__dot" part="dot"></div>
    <div class="ld-radio__box" part="box"></div>
  </mock:shadow-root>
</ld-radio>
`;

exports[`ld-radio heighlight 1`] = `
<ld-radio class="ld-radio ld-radio--highlight" mode="highlight" part="root">
  <mock:shadow-root>
    <input part="input focusable" tabindex="-1" type="radio">
    <div class="ld-radio__dot" part="dot"></div>
    <div class="ld-radio__box" part="box"></div>
  </mock:shadow-root>
</ld-radio>
`;

exports[`ld-radio renders 1`] = `
<ld-radio class="ld-radio" part="root">
  <mock:shadow-root>
    <input part="input focusable" tabindex="-1" type="radio">
    <div class="ld-radio__dot" part="dot"></div>
    <div class="ld-radio__box" part="box"></div>
  </mock:shadow-root>
</ld-radio>
`;

exports[`ld-radio tone 1`] = `
<ld-radio class="ld-radio ld-radio--dark" part="root" tone="dark">
  <mock:shadow-root>
    <input part="input focusable" tabindex="-1" type="radio">
    <div class="ld-radio__dot" part="dot"></div>
    <div class="ld-radio__box" part="box"></div>
  </mock:shadow-root>
</ld-radio>
`;
